brief-history.tex

```diff
@@ -931,77 +931,124 @@ \subsection{GraphBLAS}
+performance for primary use cases.
 
 We mentioned ``batched'' linear algebra interfaces briefly above.
 These perform many small dense operations in an optimized way,
-generally with the intention of parallelization
-\cite{dongarra2016batched}.  Different vendors (e.g.,
-Intel\footnote{See
+generally with the intention of parallelization and/or vectorization
+\cite{dongarra2016batched}.  Many applications that use dense linear
+algebra actually do many small, often independent dense operations.
+Scientific and engineering computing have plenty of examples, but the
+strong and growing interest in machine learning also calls for batched
+linear algebra.
+
+The interface of batched linear algebra operations matters a lot for
+performance, but may constrain generality.  For example, requiring a
+specific data layout and constraining all matrices to have the same
+dimensions makes vectorization easier, but applications in sparse
+multifrontal matrix factorizations may produce dense matrices of
+different dimensions.  Vendors like Intel\footnote{See
   \url{https://software.intel.com/en-us/articles/introducing-batch-gemm-operations}.}
 and NVIDIA\footnote{See
-  \url{https://docs.nvidia.com/cuda/cublas/index.html}.}) have
-different interfaces, and the interface matters a lot for performance.
-Relton et al.\ survey different interface options
-\cite{relton2016comparison}.
+  \url{https://docs.nvidia.com/cuda/cublas/index.html}.} have
+different interfaces with different levels of generality.  Relton et
+al.\ survey different interface options \cite{relton2016comparison}.
+A standardization effort began circa 2016,\footnote{See
+  \cite{dongarra2016batched} and \url{http://icl.utk.edu/bblas/}.}
+and continues with regular workshops and publications.
 
 The \texttt{mdspan} \cite{P0009r8} multidimensional array data
 structure could give users a way to store a batch of small dense
 matrices in an optimized layout.  It also could give a linear algebra
-library a way to express interface polymorphism.
+library a way to express interface polymorphism.  For example,
+\texttt{mdspan} supports different layouts.  An implementation could
+provide generic algorithms, but specialize for specific layouts.  Some
+layouts can express different ways to pack linear algebra objects that
+can help with vectorization.
 
 \section{Lessons learned from Python libraries}
 \label{S:Python}
 
-Scientists saw Python in the 2000s as an alternative for \texttt{Matlab}, itself
- an alternative interpreted langage to Fortran. Despite not having native
-  support for multidimensional arrays, two main projects, \texttt{Numarray} and
-  \texttt{Numeric}, were merged by Travis Oliphant to create
-  \texttt{Numpy}~\cite{oliphant2006numpy}. The communities behind
-  the two other projects soon merged.
+Summary: The Python programming language does not come with native
+linear algebra or multidimensional array support, but Numpy offers a
```

Comment:
Actually, it does (see https://docs.python.org/3/c-api/buffer.html). Numpy builds now on top of it.

Reply:
@mbrucher Should we elaborate this point, or just say something like the following: "Python did not come with native multidimensional array support until version 3 of the language, but Numpy..."?

Reply:
@mbrucher Buffer is a protocol for Python's C API.  Python 3 doesn't have multidimensional arrays in the Python language, as far as I can tell.  Is that right?

Reply:
I guess it is. It's a little bit fuzzy I have to say, as this C-API compatibility is central now for lots of array types (numpy, pillow, TF...)

Reply:
@mbrucher Thanks for the comment! :-) While I use Python sometimes, I'm not an expert.  Would you like to write this bit, or should I just try?  I could add what you just said, that many Python libraries use the Buffer protocol to implement array types.

Reply:
I'll try to write something tomorrow, I have to take some time on SG19 paper, I'll do both at the same time. Thanks a lot for everything you already did!